attachments/04_logical_device.cpp

```diff
@@ -191,25 +192,45 @@ class HelloTriangleApplication {
         // get the first index into queueFamilyProperties which supports graphics
         auto graphicsQueueFamilyProperty = std::ranges::find_if( queueFamilyProperties, []( auto const & qfp )
                         { return (qfp.queueFlags & vk::QueueFlagBits::eGraphics) != static_cast<vk::QueueFlags>(0); } );
+        assert(graphicsQueueFamilyProperty != queueFamilyProperties.end() && "No graphics queue family found!");
 
         auto graphicsIndex = static_cast<uint32_t>( std::distance( queueFamilyProperties.begin(), graphicsQueueFamilyProperty ) );
 
+#if 1
         // query for Vulkan 1.3 features
-        auto features = physicalDevice.getFeatures2();
-        vk::PhysicalDeviceVulkan13Features vulkan13Features;
-        vk::PhysicalDeviceExtendedDynamicStateFeaturesEXT extendedDynamicStateFeatures;
-        vulkan13Features.dynamicRendering = vk::True;
-        extendedDynamicStateFeatures.extendedDynamicState = vk::True;
-        vulkan13Features.pNext = &extendedDynamicStateFeatures;
-        features.pNext = &vulkan13Features;
+        vk::StructureChain<vk::PhysicalDeviceFeatures2, vk::PhysicalDeviceVulkan13Features, vk::PhysicalDeviceExtendedDynamicStateFeaturesEXT> featureChain = {
+            {},                               // vk::PhysicalDeviceFeatures2
+            {.dynamicRendering = true },      // vk::PhysicalDeviceVulkan13Features
+            {.extendedDynamicState = true }   // vk::PhysicalDeviceExtendedDynamicStateFeaturesEXT
+        };
+
```

Comment:
That's really cool, I didn't even know this was available.  We should also find a place in the tutorial text to call this out, describe pros/cons and recommend it.  If you'd like go ahead and add this to the rest of the chapters and I'm happy to draft the language in the tutorial itself unless you wanna do that?

Reply:
> If you'd like go ahead and add this to the rest of the chapters

With "this", you mean the approach right above?
I also like that one a little more, even though it needs that `vk::PhysicalDeviceFeatures2` as an anchor for the structs, which isn't used besides that.
The other approach doesn't need that struct, the feature structs are simple chained to `vk::DeviceCreateInfo`.

> I'm happy to draft the language in the tutorial itself

Sounds good.

Reply:
Yep, i meant the first approach.  For some reason to my eye it still looks cleaner even if it requires an extra struct rather than directly to the device create info.  Maybe because it feels closer to what I'm used to with the C version to use the extra struct or at least kinda follow what I'm expecting.